win_chocolatey - TLS Installing with restricted TLS

<!---
Verify first that your issue/request is not already reported on GitHub.
THIS FORM WILL BE READ BY A MACHINE, COMPLETE ALL SECTIONS AS DESCRIBED.
Also test if the latest release, and devel branch are affected too.
ALWAYS add information AFTER (OUTSIDE) these html comments.
Otherwise it may end up being automatically closed by our bot. -->

##### SUMMARY
win_chocolatey fails on initial install due to https://github.com/chocolatey/choco/wiki/Installation#installing-with-restricted-tls. 

Returns error:
```
"module_stderr": "An error occurred while creating the pipeline.\r\n    + CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordE \r\n   xception\r\n    + FullyQualifiedErrorId : RuntimeException\r\n \r\nChocolatey-Install-Upgrade : Exception calling \"DownloadString\" with \"1\" \r\nargument(s): \"The underlying connection was closed: An unexpected error \r\noccurred on a receive.\"\r\nAt line:516 char:1\r\n+ Chocolatey-Install-Upgrade\r\n+ ~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : NotSpecified: (:) [Chocolatey-Install-Upgrade],  \r\n   MethodInvocationException\r\n    + FullyQualifiedErrorId : WebException,Chocolatey-Install-Upgrade\r\n \r\n\r\n",
```

```
An error occurred while creating the pipeline.
    + CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordE
   xception
    + FullyQualifiedErrorId : RuntimeException

Chocolatey-Install-Upgrade : Exception calling \"DownloadString\" with \"1\"
argument(s): \"The underlying connection was closed: An unexpected error
occurred on a receive.\"
At line:516 char:1
+ Chocolatey-Install-Upgrade
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Chocolatey-Install-Upgrade],
   MethodInvocationException
    + FullyQualifiedErrorId : WebException,Chocolatey-Install-Upgrade
```


##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
<!--- Insert, BELOW THIS COMMENT, the name of the module, plugin, task or feature.
Do not include extra details here, e.g. "vyos_command" not "the network module vyos_command" or the full path-->
win_chocolatey

##### ANSIBLE VERSION
<!--- Paste, BELOW THIS COMMENT, verbatim output from "ansible --version" between quotes below -->
```
2.4.2.0
```

##### CONFIGURATION
<!--- If using Ansible 2.4 or above, paste, BELOW THIS COMMENT, the results of "ansible-config dump --only-changed"
Otherwise, mention any settings you have changed/added/removed in ansible.cfg
(or using the ANSIBLE_* environment variables).-->
Ansible 2.4.2.0 
---
```
ansible_winrm_server_cert_validation: ignore
ansible_connection: winrm
ansible_user: --
ansible_winrm_scheme: https
ansible_port: 5986
ansible_winrm_transport: kerberos
```
##### OS / ENVIRONMENT
<!--- Mention, BELOW THIS COMMENT, the OS you are running Ansible from, and the OS you are
managing, or say "N/A" for anything that is not platform-specific.
Also mention the specific version of what you are trying to control,
e.g. if this is a network bug the version of firmware on the network device.-->
From RHEL7 to Windows Server 2012.

##### STEPS TO REPRODUCE
<!--- For bugs, show exactly how to reproduce the problem, using a minimal test-case.
For new features, show how the feature would be used. -->

<!--- Paste example playbooks or commands between quotes below -->
```yaml

```

<!--- You can also paste gist.github.com links for larger files -->

##### EXPECTED RESULTS
<!--- What did you expect to happen when running the steps above? -->

##### ACTUAL RESULTS
<!--- What actually happened? If possible run with extra verbosity (-vvvv) -->

<!--- Paste verbatim command output between quotes below -->
```

```

Unfortunately I am a bit stuck trying to get this setup on my local before I test the ansible update to tower. Sorry on the slow progress on this.

Looking through a few older threads on how to get kerberos for python/windows setup working on my rhel7 virtualbox but getting stuck here with the gcc, kerberos libs, etc. dependencies -- mainly now with pykerberos. I think once I get this I can alias or something on tower server to get it working w/ the ansible .git from your jbordean93 branch.

After doing a pip install git+https://github.com/jborean93/ansible.git@chocolatey-tls1.2


See below.

Installing collected packages: pykerberos, requests-kerberos
  Running setup.py install for pykerberos ... error
    Complete output from command /opt/rh/python27/root/usr/bin/python2 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-30d3if/pykerberos/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-xr4DLO/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_ext
    building 'kerberos' extension
    creating build
    creating build/temp.linux-x86_64-2.7
    creating build/temp.linux-x86_64-2.7/src
    gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -I/opt/rh/python27/root/usr/include/python2.7 -c src/kerberos.c -o build/temp.linux-x86_64-2.7/src/kerberos.o -DGSSAPI_EXT
    src/kerberos.c:17:20: fatal error: Python.h: No such file or directory
     #include <Python.h>
                        ^
    compilation terminated.
    error: command 'gcc' failed with exit status 1
    
    ----------------------------------------
Command "/opt/rh/python27/root/usr/bin/python2 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-30d3if/pykerberos/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-xr4DLO/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-30d3if/pykerberos/
[root@john-redhat ~]#

--- a/changelogs/fragments/win_chocoalatey_install_tls12.yml
+++ b/changelogs/fragments/win_chocoalatey_install_tls12.yml
@@ -1,0 +1,2 @@
+bugfixes:
+- win_chocolatey - enable TLSv1.2 support when downloading the Chocolatey installer https://github.com/ansible/ansible/issues/41906

--- a/lib/ansible/modules/windows/win_chocolatey.ps1
+++ b/lib/ansible/modules/windows/win_chocolatey.ps1
@@ -48,8 +48,17 @@ Function Chocolatey-Install-Upgrade
     $ChocoAlreadyInstalled = Get-Command -Name "choco.exe" -ErrorAction SilentlyContinue
     if ($ChocoAlreadyInstalled -eq $null)
     {
+        # We need to install chocolatey
+        # Enable TLS1.1/TLS1.2 if they're available but disabled (eg. .NET 4.5)
+        $security_protcols = [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::SystemDefault
+        if ([Net.SecurityProtocolType].GetMember("Tls11").Count -gt 0) {
+            $security_protcols = $security_protcols -bor [Net.SecurityProtocolType]::Tls11
+        }
+        if ([Net.SecurityProtocolType].GetMember("Tls12").Count -gt 0) {
+            $security_protcols = $security_protcols -bor [Net.SecurityProtocolType]::Tls12
+        }
+        [Net.ServicePointManager]::SecurityProtocol = $security_protcols
 
-        #We need to install chocolatey
         $wc = New-Object System.Net.WebClient;
         if ($proxy_url)
         {

